File: squint/rust/rosetta-squint/examples/squint-cli.rs
//! squint-cli — compute a perceptual image hash and print the hex string to stdout.
//!
//! Used by tools/cross-squint-diff for live cross-port equivalence checking.
//!
//! Usage: squint-cli <algo> <size> <path>
//!   algo  — one of: phash, phash_simple, dhash, dhash_vertical, average_hash,
//!           whash_haar, whash_db4, whash_db4_robust, colorhash, crop_resistant_hash
//!   size  — hash_size (or binbits for colorhash); pass "-" for crop_resistant_hash
//!   path  — path to the image file
//!
//! Exit:  0 on success, 1 on hash/decode error, 2 on usage error.

use std::env;
use std::process::ExitCode;

use rosetta_squint::{
    average_hash, colorhash, crop_resistant_hash, dhash, dhash_vertical, phash, phash_simple,
    whash_db4, whash_db4_robust, whash_haar,
};

fn main() -> ExitCode {
    let args: Vec<String> = env::args().collect();
    if args.len() != 4 {
        eprintln!("usage: squint-cli <algo> <size> <path>");
        return ExitCode::from(2);
    }
    let algo = args[1].as_str();
    let size: usize = args[2].parse().unwrap_or(8);
    let path = &args[3];

    let result = match algo {
        "phash" => phash(path, size).map(|h| h.to_hex()),
        "phash_simple" => phash_simple(path, size).map(|h| h.to_hex()),
        "dhash" => dhash(path, size).map(|h| h.to_hex()),
        "dhash_vertical" => dhash_vertical(path, size).map(|h| h.to_hex()),
        "average_hash" => average_hash(path, size).map(|h| h.to_hex()),
        "whash_haar" => whash_haar(path, size).map(|h| h.to_hex()),
        "whash_db4" => whash_db4(path, size).map(|h| h.to_hex()),
        "whash_db4_robust" => whash_db4_robust(path, size).map(|h| h.to_hex()),
        "colorhash" => colorhash(path, size).map(|h| h.to_hex()),
        "crop_resistant_hash" => crop_resistant_hash(path, None).map(|mh| mh.to_hex()),
        _ => {
            eprintln!("unknown algo: {}", algo);
            return ExitCode::from(2);
        }
    };

    match result {
        Ok(hex) => {
            println!("{}", hex);
            ExitCode::from(0)
        }
        Err(e) => {
            eprintln!("error: {}", e);
            ExitCode::from(1)
        }
    }
}

File: squint/java/rosetta-squint/src/main/java/io/github/wmetcalf/rosettasquint/cli/SquintCli.java
package io.github.wmetcalf.rosettasquint.cli;

import io.github.wmetcalf.rosettasquint.Squint;

/**
 * squint-cli — compute a perceptual image hash and print the hex string to stdout.
 *
 * <p>Used by tools/cross-squint-diff for live cross-port equivalence checking.
 *
 * <p>Usage: squint-cli &lt;algo&gt; &lt;size&gt; &lt;path&gt;
 * <ul>
 *   <li>algo  — one of: phash, phash_simple, dhash, dhash_vertical, average_hash,
 *               whash_haar, whash_db4, whash_db4_robust, colorhash, crop_resistant_hash</li>
 *   <li>size  — hash_size (or binbits for colorhash); pass "-" for crop_resistant_hash</li>
 *   <li>path  — path to the image file</li>
 * </ul>
 *
 * <p>Exit: 0 on success, 1 on hash/decode error, 2 on usage error.
 */
public final class SquintCli {
    private SquintCli() {}

    public static void main(String[] args) {
        if (args.length != 3) {
            System.err.println("usage: squint-cli <algo> <size> <path>");
            System.exit(2);
        }
        String algo = args[0];
        int size = 8;
        try {
            size = Integer.parseInt(args[1]);
        } catch (NumberFormatException ignored) {
            // "-" placeholder for crop_resistant_hash — size stays 8 (unused)
        }
        String path = args[2];

        String hex;
        try {
            hex = switch (algo) {
                case "phash"              -> Squint.phash(path, size).toString();
                case "phash_simple"       -> Squint.phashSimple(path, size).toString();
                case "dhash"              -> Squint.dhash(path, size).toString();
                case "dhash_vertical"     -> Squint.dhashVertical(path, size).toString();
                case "average_hash"       -> Squint.averageHash(path, size).toString();
                case "whash_haar"         -> Squint.whashHaar(path, size).toString();
                case "whash_db4"          -> Squint.whashDb4(path, size).toString();
                case "whash_db4_robust"   -> Squint.whashDb4Robust(path, size).toString();
                case "colorhash"          -> Squint.colorhash(path, size).toString();
                case "crop_resistant_hash"-> Squint.cropResistantHash(path).toString();
                default -> {
                    System.err.println("unknown algo: " + algo);
                    System.exit(2);
                    yield "";
                }
            };
        } catch (Exception e) {
            System.err.println("error: " + e.getMessage());
            System.exit(1);
            return;
        }
        System.out.println(hex);
    }
}
